Add Gulpfile.ts

weswigham opened the pull request:
Last fall I always said I'd probably get around to doing this at some point.

I was frustrated earlier today by our Jake-based build, and I had always talked about how it'd be easier to work with as another task runner. So this evening I did just what I always said I would - I ported our Jakefile to a Gulpfile. Added bonus: It's now written in typescript! And there's a `gulp help` command!

Gulp is orders of magnitude more popular than Jake nowadays, and sees much more active use - external contributors are more likely to have existing experiences with gulp. We can also continue moving parts of the build to in-memory streams for a bit more efficiency once this is done. 

@DanielRosenwasser @sandersn @yuit You guys should definitely take a look.

sandersn commented:
@rbuckton can you take a look too? I think you know the most about gulp around here.

sandersn commented:
I almost always C-c tests when they pass and start linting. And when I realise I'm building the wrong thing, etc. The rename trick probably isn't going to work. :(

sandersn commented on Jakefile.js:
```diff
@@ -1,1076 +0,0 @@
-var LKGDirectory = "lib/";
-
-var copyright = "CopyrightNotice.txt";
-var thirdParty = "ThirdPartyNoticeText.txt";
```
I don't see this (yet) in the gulpfile.

weswigham replied:
It was unused. I'll double check if its actually used in the Jakefile and that I forgot something.

sandersn replied:
I double checked now that I am not on the bus. It's unused.

sandersn commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1083 @@
+}
+
+function needsUpdate(source: string | string[], dest: string | string[]): boolean {
+    if (typeof source === "string" && typeof dest === "string") {
```
seems like this could be refactored with some comparison function *cmp* into 4 cases: 
```
  | cmp(src,dst) 
  | every(cmp(src), dst) 
  | every(flip(cmp)(dst), src) 
  | every(uncurry(cmp), zip(src, dst))
where cmp src dst = ...
```

weswigham replied:
Originally I didn't want to write the functional utility functions... But now that its not super late at night I realize that I can literally just include them from somewhere, since this is just our build, or write them in another file, or even include them from compiler utils. So yeah, I agree.

sandersn replied:
You don't necessarily need `flip`, `zip`, or `uncurry`. `needsUpdate` would be better even with adapter lambdas that do the same thing.

Well, you might need `zip` or `zipWith`.

yuit commented:
talk with @sandersn instead of deleting the JakeFile completely, may be we should keep the JakeFile for now and travis-CI to use Jake. In the team, we will start using gulp internally for a week or so before switch. Also @sandersn suggests that we should do this on Mon/Tue. instead of Friday. if we decide to merge so things won't break on weekend :smile_cat:

rbuckton commented:
Have you considered using the `gulp-typescript` plugin to replace `compileFile` and leverage our existing `tsconfig.json` files?

weswigham commented:
@yuit  Considering its build is still broken, I have no illusions that this is getting merged this weekend, lol.

@rbuckton Yes! This is a thing I want to do! The only reason I didn't was because I wanted to start to replicating our Jakefile's functionality as closely as possible. (Maybe we wanted to keep on using our command-line interface? I dunno)

weswigham commented:
@sandersn I should have a fix incoming for the terribad workaround - passing an empty array to the `types` compiler option (thanks @RyanCavanaugh). Unfortunately, we don't (presently) support passing an empty array on the command line - so I'm going to make a PR for that. Then I'm going to swap us to using `gulp-typescript` in the gulpfile in this branch, which should also remove the issue.

sandersn commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1084 @@
+    });
+});
+
+gulp.task("generate-code-coverage", "Generates code coverage data via instanbul", ["tests"], (done) => {
```
Typo: istanbul

sandersn commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1084 @@
+                    if (err) fail(err);
+                    prependFile(copyright, standaloneDefinitionsFile);
+
+                    // Stanalone/web definition file using global 'ts' namespace
```
Typo:standalone

sandersn commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1084 @@
+function getDiffTool() {
+    const program = process.env["DIFF"];
+    if (!program) {
+        console.error("Add the 'DIFF' environment constiable to the path of the program you want to use.");
```
Typo: variable instead of constiable

sandersn replied:
And elsewhere below

sandersn commented:
I vote that you undo the removal of the Jakefile and make sure Travis still works with jake, then merge this. We can test it over the next week or so and then switch Travis.

weswigham commented:
@sandersn @rbuckton  @yuit  Updated! Kept the Jakefile, as requested (and the package.json pointing at jake), and rewrote all the compilation tasks in the gulpfile to use `gulp-typescript` which, conveniently, removed the need for the `@types` folder hackery. Another interesting change this required was that instead of just building tsc.js to use the built local compiler, the entire services layer has to get made instead (in order to make our node-compatible typescript.js).

sandersn commented:
:+1:

weswigham commented:
I need to run an LKG to get the empty-list compiler options parsing into the lib folder in order to prevent the Jakefile from breaking. @sandersn @DanielRosenwasser should an LKG be done in another PR or as part of this one?

weswigham commented:
Eh, did an LKG here. I can revert it if that's undesirable.

weswigham commented:
@rbuckton Do you want to take another look at this, since the gulpfile's changed significantly (namely, no longer shelling out to compile).

rbuckton commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1125 @@
+
+const cmdLineOptions = minimist(process.argv.slice(2), {
+    boolean: ["debug", "light", "colors", "lint", "soft"],
+    string: ["browser", "tests", "host", "reporter", ],
```
nit: remove trailing comma

rbuckton commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1125 @@
+
+function exec(cmd: string, args: string[], complete: () => void = (() => {}), error: (e: any, status: number) => void = (() => {})) {
+    console.log(`${cmd} ${args.join(" ")}`);
+    const ex = cp.spawn(cmd, args);
```
consider:
```ts
const ex = cp.spawn(cmd, args, { stdio: "inherit" });
```
Then you can drop the `ex.stdout.pipe(process.stdout);` and `ex.stderr.pipe(process.stderr);` lines.

rbuckton replied:
Also if you do:
```ts
const ex = cp.spawn(win ? "cmd" : "/bin/sh", [win ? "/c" : "-c", cmd, ...args], { stdio: "inherit", windowsVerbatimArguments: true });
```
It will use the OS shell to launch the command, letting `spawn` leverage the `PATH` environment variable and automatically handle executables without appending `.cmd`, etc.
```

rbuckton commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1125 @@
+const LKGCompiler = path.join(LKGDirectory, compilerFilename);
+const builtLocalCompiler = path.join(builtLocalDirectory, compilerFilename);
+
+// add node_modules to path so we don"t need global modules, prefer the modules by adding them first
```
This comment no longer accurately reflects what is happening in the following lines.

weswigham replied:
Truth. Will remove.

rbuckton commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1125 @@
+            return false;
+        }
+    }
+    else if (typeof source !== "string" && typeof dest !== "string") {
```
I don't see a call site that uses this case, but if there was do you expect it to compare files on a one to one basis?

weswigham replied:
Yes.

rbuckton commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1125 @@
+const packageJson = "package.json";
+const programTs = path.join(compilerDirectory, "program.ts");
+
+function needsUpdate(source: string | string[], dest: string | string[]): boolean {
```
Have you considered leveraging vinyl streams to handle this? `vinyl-fs` already stats the files as they are matched.

rbuckton replied:
Consider this: https://github.com/gulpjs/gulp/blob/master/docs/recipes/only-pass-through-changed-files.md

weswigham replied:
The only place it's in use where streams actually get used is for concatenating the lib (everywhere else it's used before shelling out). However, I could probably change the lib creation tasks to use `gulp-newer` or `gulp-changed`.

rbuckton commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1125 @@
+});
+
+gulp.task(servicesFile, false, ["lib", "generate-diagnostics"], (done) => {
+    const settings: tsc.Settings = getCompilerSettings({
```
Why can't we use `tsc.createProject` here?

weswigham replied:
We probably can.

rbuckton commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1125 @@
+    if (!useDebugMode) {
+        result = result.pipe(insert.prepend(fs.readFileSync(copyright)));
+    }
+    result.pipe(sourcemaps.write("."))
```
Does this properly set `sourceRoot`? In my experience `gulp-sourcemaps` doesn't usually include the source root unless its specified in an option. Have you verified you can debug the compiler after it is built?

rbuckton replied:
There might be a better way to structure this so that you can `return` something gulp can monitor rather than relying on `console.error`. 

Consider using something like [merge2](https://www.npmjs.com/package/merge2) to merge the events from multiple streams to report errors back to gulp.

weswigham replied:
I had actually been merging them originally but I had some random issue with errors getting swallowed. I probably need to merge in the sequence of streams so that errors get merged at all stages?

rbuckton commented on Gulpfile.ts:
```diff
@@ -0,0 +1,1125 @@
+    let result: NodeJS.ReadWriteStream = localCompilerProject.src()
+        .pipe(sourcemaps.init())
+        .pipe(tsc(localCompilerProject));
+    if (!useDebugMode) {
```
Consider using [gulp-if](https://www.npmjs.com/package/gulp-if)

weswigham commented:
@rbuckton Updated to handle your comments and, now that `newer` is guarding every stream, add a `watch` task which runs `runtests-parallel`.

weswigham commented:
Sourcemaps should now work in every context, though I think I can improve the `runtests-browser` ones a bit.

rbuckton commented on Gulpfile.ts:
```diff
@@ -582,25 +508,16 @@ gulp.task(tsserverLibraryFile, false, [servicesFile], (done) => {
-    js.pipe(sourcemaps.write("."))
+
+    return merge2([
+        js.pipe(gIf(useDebugMode, insert.prepend(fs.readFileSync(copyright))))
```
Can we make this less eager using `lazypipe`?

rbuckton commented on Gulpfile.ts:
```diff
@@ -434,55 +435,50 @@ const nodePackageFile = path.join(builtLocalDirectory, "typescript.js");
         .pipe(sourcemaps.init())
         .pipe(tsc(localCompilerProject))
-        .pipe(gIf(useDebugMode, insert.prepend(fs.readFileSync(copyright))))
+        .pipe(gIf(useDebugMode, prependCopyright()))
```
I played around with this and this still ends up eagerly calling `fs.readFileSync`.

Another option that does not use `gulp-if`/`lazypipe` would be:

```ts
import { PassThrough } from "stream";

...

    return localCompilerProject.src()
        .pipe(newer(builtLocalCompiler))
        .pipe(sourcemaps.init())
        .pipe(tsc(localCompilerProject)) 
        .pipe(useDebugMode ? insert.prepend(fs.readFileSync(copyright)) : pass())
        .pipe(sourcemaps.write(".")) 
        .pipe(gulp.dest(builtLocalDirectory));

...

function pass() {
  return new PassThrough({ readableObjectMode: true, writableObjectMode: true });
}
```

weswigham replied:
Or why not just `insert.prepend(useDebugMode ? fs.readFileSync(copyright) : "")`, any avoid both of those dependencies?

rbuckton commented on Gulpfile.ts:
```diff
@@ -748,7 +745,7 @@ gulp.task("runtests",
 const nodeServerInFile = "tests/webTestServer.ts";
 gulp.task(nodeServerOutFile, false, [servicesFile], () => {
-    const settings: tsc.Settings = getCompilerSettings({}, /*useBuiltCompiler*/ true);
+    const settings: tsc.Settings = getCompilerSettings({module: "commonjs"}, /*useBuiltCompiler*/ true);
```
nit: whitespace

weswigham replied:
I'm not sure what whitespace you're complaining about, but whatever it is I guess it should be codified because our lint rules don't catch it.

sandersn replied:
probably `{ module: "commonjs" }` -- spaces after `{` and before`}`

weswigham replied:
Yeah - this style is inconsistently enforced, I have instances both with and without spaces in the compiler. We need a lint rule for this, one way or the other, if people are going to display a preference.

weswigham replied:
Additionally, is this supposed to be consistent with binding patterns and/or template strings? I'm internally consistent in the Gulpefile (no leading/trailing whitespace for any), but again, we're inconsistent within the compiler.

sandersn replied:
I'm not sure off the top of my head. I "read somewhere" that the Javascript standard syntax was with extra spaces, but I don't remember how *actually* authoritative the source was. I'll try to track down the source.

weswigham replied:
TBH, the lack of leading/trailing whitespace is, for me, one of the things which makes an object literal or binding pattern distinct from a function or lambda body.

sandersn replied:
I can't find it, so probably we should just have a discussion and vote on which linter rule we want to turn on. (Does TSLint have it? ESLint does -- [object curly spacing](http://eslint.org/docs/rules/object-curly-spacing))

weswigham replied:
I don't see any tslint rules white would cover whitespace before/after curlys.

weswigham replied:
I've written a rule to check which is the dominant style - by and large with whitespace is more prevalent in our codebase, so I assume the default should be with whitespace.

weswigham replied:
I'm going to merge this without fixing these nits and the fix them when I merge the lint rule catching them, is that okay?

sandersn replied:
Works for me. @rbuckton ?

rbuckton commented on Gulpfile.ts:
```diff
@@ -970,7 +967,7 @@ const tslintRulesFiles = tslintRules.map(function(p) {
     const pathname = path.join(builtLocalDirectory, "tslint", p + ".js");
     gulp.task(pathname, false, [], () => {
-        const settings: tsc.Settings = getCompilerSettings({}, /*useBuiltCompiler*/ false);
+        const settings: tsc.Settings = getCompilerSettings({module: "commonjs"}, /*useBuiltCompiler*/ false);
```
nit: whitespace

Do we need to start linting our gulpfile? 😄

rbuckton commented on Gulpfile.ts:
```diff
@@ -27,7 +27,7 @@ declare global {
 import mkdirP = require("mkdirP");
 import minimist = require("minimist");
-import browserify = require("browserify");
+import browserify = require("browserify")
```
nit: trailing semicolon.

weswigham replied:
We are linting the gulpfile if you do gulp lint. I should update the
jakefile to lint it, too.

On Wed, Jun 22, 2016, 6:42 PM Ron Buckton <notifications@github.com> wrote:

> In Gulpfile.ts
> <https://github.com/Microsoft/TypeScript/pull/9068#discussion_r68164578>:
>
> > @@ -27,7 +27,7 @@ declare global {
> >  import del = require("del");
> >  import mkdirP = require("mkdirP");
> >  import minimist = require("minimist");
> > -import browserify = require("browserify");
> > +import browserify = require("browserify")
>
> nit: trailing semicolon.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Microsoft/TypeScript/pull/9068/files/8aeb682206d4eefffadd2b7de6a614a2cb88c1a6..d008da5888181a6f15823b876c7daa391049c16e#r68164578>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/ACzAMuFsOQZ36AA_aGh8JPbvdd9tZITmks5qOeShgaJpZM4Iyq8C>
> .
>
